app/api/common.v2/database/MongoDataSource.ts

```diff
@@ -1,28 +1,132 @@
-import { Db, Document } from 'mongodb';
+import { Collection, Db, Document } from 'mongodb';
 import { BulkWriteStream } from './BulkWriteStream';
 import { MongoTransactionManager } from './MongoTransactionManager';
 
+type MaxAmountOfParameters<F extends (...args: any) => any> = Required<Parameters<F>>['length'];
+
+/**
+ * Map of collection members to position of its options argument.
+ * Position is based on 1.
+ */
+type MethodsOptionsArgPosition = {
+  [method in keyof Collection]: Collection[method] extends (...args: any) => any
+    ? MaxAmountOfParameters<Collection[method]>
+    : null;
+};
+
 export abstract class MongoDataSource<CollectionSchema extends Document = any> {
   protected db: Db;
 
   protected abstract collectionName: string;
 
   private transactionManager: MongoTransactionManager;
 
+  private collectionProxy?: Collection<CollectionSchema>;
+
   constructor(db: Db, transactionManager: MongoTransactionManager) {
     this.db = db;
     this.transactionManager = transactionManager;
   }
 
+  static scopedMethods: MethodsOptionsArgPosition = {
+    insertOne: 2,
+    insertMany: 2,
+    bulkWrite: 2,
+    updateOne: 3,
+    replaceOne: 3,
+    updateMany: 3,
+    deleteOne: 2,
+    deleteMany: 2,
+    rename: 2,
+    drop: 1,
+    findOne: 2,
+    find: 2,
+    options: 1,
+    isCapped: 1,
+    createIndex: 2,
+    createIndexes: 2,
+    dropIndex: 2,
+    dropIndexes: 1,
+    listIndexes: 1,
+    indexExists: 2,
+    indexInformation: 1,
+    estimatedDocumentCount: 1,
+    countDocuments: 2,
+    distinct: 3,
+    indexes: 1,
+    stats: 1,
+    findOneAndDelete: 2,
+    findOneAndReplace: 3,
+    findOneAndUpdate: 3,
+    aggregate: 2,
+    watch: 2,
+    initializeUnorderedBulkOp: 1,
+    initializeOrderedBulkOp: 1,
+    count: 2,
+    dbName: null,
```

Comment:
Cause I couldn't figure out the typings. It doesn't seem to be a way to remove attributes from an interface based on the type of its value.
Typing them undefined instead of null won't make them optional, you'll still have to include them with undefined as value.
Typing them as never, will fail in all cases, cause you need to include them since they're not optional, but as you include it there's no value of type never, so it will also fail.
The apparent solution is to make them never and optional, but there's no way to only make some properties optional in a mapped type. So... :)